File: db4ounit/src/com/db4o/test/unit/TestCase.java
package com.db4o.test.unit;

public abstract class TestCase extends Assert implements Test {
	
	public void run(TestResult result) {
		try {
			setUp();
			run();
		}
		catch(Exception exc) {
			result.fail(exc);
		}
		finally {
			try {
				tearDown();
			} catch (Exception exc) {
				result.fail(exc);
			}
		}
	}
	
	protected void setUp() throws Exception {}

	protected void tearDown() throws Exception {}

	protected abstract void run() throws Exception;
}

File: db4on/instrumentation/Db4oAdmin.Tests/Framework/TestCase.cs
using System.Collections;
using System.Reflection;

namespace Db4oAdmin.Tests.Framework
{
	public class TestMethod : ITest
	{
		private TestCase _testCase;
		private MethodInfo _method;

		public TestMethod(TestCase testCase, MethodInfo method)
		{
			_testCase = testCase;
			_method = method;
		}
		
		public string Name
		{
			get { return _method.Name; }
		}
		
		public void Run()
		{
			try
			{
				_testCase.SetUp();
				_method.Invoke(_testCase, null);
			}
			finally
			{
				_testCase.TearDown();
			}
		}
	}
	
	public abstract class TestCase : ITestSuiteBuilder
	{
		public virtual void SetUp()
		{
		}

		public virtual void TearDown()
		{
		}

		public ITest[] Build()
		{
			ArrayList tests = new ArrayList();
			foreach (MethodInfo method in GetType().GetMethods(BindingFlags.Public | BindingFlags.Instance))
			{
				if (!method.Name.StartsWith("Test")) continue;
				tests.Add(new TestMethod(this, method));
			}
			return (ITest[])tests.ToArray(typeof (ITest));
		}
	}
}